In [2]:
import os
from dotenv import load_dotenv
from autogen_agentchat.agents import AssistantAgent
from autogen_ext.models.openai import OpenAIChatCompletionClient
from autogen_agentchat.teams import RoundRobinGroupChat
from autogen_agentchat.ui import Console

load_dotenv()
or_key = os.getenv("OPENROUTER_API_KEY")

In [14]:
# 1. The Cloud Brain (Auto-Routing Load Balancer)
# Since this is a text task, the router will dynamically bypass all 429 traffic jams!
cloud_client = OpenAIChatCompletionClient(
    model="openrouter/free", 
    api_key=or_key,
    base_url="https://openrouter.ai/api/v1",
    model_info={"vision": False, "function_calling": True, "json_output": False, "family": "unknown"}
)

In [15]:
# 2. The Local Brain (Fast Processing/Formatting)
local_client = OpenAIChatCompletionClient(
    model="llama3.2",
    api_key="placeholder",
    base_url="http://localhost:11434/v1",
    model_info={"vision": False, "function_calling": True, "json_output": False, "family": "llama3"}
)

print("✅ Block 1 & 2: Cloud Brain swapped to the OpenRouter Free Load-Balancer!")

✅ Block 1 & 2: Cloud Brain swapped to the OpenRouter Free Load-Balancer!


In [16]:
# 3a. The Cloud Researcher (Powered by Llama 3.3 70B)
cloud_researcher = AssistantAgent(
    name="Cloud_Researcher",
    model_client=cloud_client,
    system_message="You are a brilliant agricultural researcher. When asked a question, provide a detailed, 2-paragraph analysis."
)

# 3b. The Local Formatter (Powered by Local Llama 3.2)
local_formatter = AssistantAgent(
    name="Local_Formatter",
    model_client=local_client,
    system_message="You are a strict data formatter. Take the researcher's output and condense it into exactly 3 concise bullet points. Do not add any conversational filler like 'Here are the points'."
)

print("✅ Block 3: Agents Created with Specialized Personas")

✅ Block 3: Agents Created with Specialized Personas


In [17]:
# 4. Create the Hybrid Team
hybrid_team = RoundRobinGroupChat(
    participants=[cloud_researcher, local_formatter],
    max_turns=2 # Turn 1: Cloud researches. Turn 2: Local formats.
)

# 5. Execute the Team Workflow (With Resilient Error Catching)
async def run_hybrid_team():
    print("--- 🚀 Starting Hybrid Multi-Agent Execution ---")
    task = "Analyze the primary benefits of precision agriculture using drone technology."
    
    try:
        # Run the stream so we can watch them talk to each other in real-time
        await Console(
            hybrid_team.run_stream(task=task)
        )
    except Exception as e:
        print(f"\n⚠️ Cloud API Congested: {e}")
        print("🔄 Wait 60 seconds and run this cell again. If it continues to fail, swap the Cloud Brain to 'google/gemma-3-27b-it:free' in Block 1.")

# THE JUPYTER FIX
await run_hybrid_team()

--- 🚀 Starting Hybrid Multi-Agent Execution ---
---------- TextMessage (user) ----------
Analyze the primary benefits of precision agriculture using drone technology.


---------- TextMessage (Cloud_Researcher) ----------
The primary benefits of precision agriculture using drone technology include enhanced crop monitoring, increased efficiency, and data-driven decision making. Drones equipped with advanced sensors can capture high-resolution images and multispectral data, allowing farmers to closely monitor crop health, identify areas of stress, and detect early signs of disease or pest infestations. This detailed information enables targeted interventions, reducing the need for blanket pesticide or fertilizer applications and promoting more sustainable farming practices. By providing a bird's-eye view of the entire field, drones help farmers identify patterns and variations in crop growth that might be missed during ground-level inspections.

Furthermore, drone technology significantly increases operational efficiency in precision agriculture. Traditional crop scouting methods are time-consuming and labor-intensive, often requiring farmers to walk or

/Users/vishalsingh/Documents/GIT/Autonomous-Research-Intelligence-Suite/venv/lib/python3.13/site-packages/autogen_agentchat/agents/_assistant_agent.py:1109: UserWarning: Resolved model mismatch: openrouter/free != arcee-ai/trinity-large-preview:free. Model mapping in autogen_ext.models.openai may be incorrect. Set the model to arcee-ai/trinity-large-preview:free to enhance token/cost estimation and suppress this warning.
  model_result = await model_client.create(


---------- TextMessage (Local_Formatter) ----------
• Precision agriculture using drone technology enhances crop monitoring through high-resolution images and multispectral data capture, enabling targeted interventions such as targeted pesticide or fertilizer applications.
• Drones significantly increase operational efficiency by allowing farmers to cover vast areas in a fraction of the time, collecting data that would be impractical to gather manually, and integrating with other precision agriculture tools.
• Data collection from drones enables data-driven decision making through real-time integration with historical information, facilitating informed decisions on planting, irrigation, harvesting, and resource optimization leading to improved productivity and profitability.


--- 🚀 Starting Hybrid Multi-Agent Execution ---
---------- TextMessage (user) ----------
Analyze the primary benefits of precision agriculture using drone technology.
---------- TextMessage (Cloud_Researcher) ----------
The primary benefits of precision agriculture using drone technology include enhanced crop monitoring, increased efficiency, and data-driven decision making. Drones equipped with advanced sensors can capture high-resolution images and multispectral data, allowing farmers to closely monitor crop health, identify areas of stress, and detect early signs of disease or pest infestations. This detailed information enables targeted interventions, reducing the need for blanket pesticide or fertilizer applications and promoting more sustainable farming practices. By providing a bird's-eye view of the entire field, drones help farmers identify patterns and variations in crop growth that might be missed during ground-level inspections.

Furthermore, drone technology significantly increases operational efficiency in precision agriculture. Traditional crop scouting methods are time-consuming and labor-intensive, often requiring farmers to walk or drive through large fields. Drones can cover vast areas in a fraction of the time, collecting data that would be impractical to gather manually. This efficiency allows for more frequent monitoring throughout the growing season, enabling farmers to respond quickly to changing conditions. The data collected by drones can be integrated with other precision agriculture tools, such as GPS-guided equipment and variable rate application systems, to optimize resource use and maximize yields. By combining real-time data with historical information, farmers can make informed decisions about planting, irrigation, and harvesting, ultimately leading to improved productivity and profitability.
/Users/vishalsingh/Documents/GIT/Autonomous-Research-Intelligence-Suite/venv/lib/python3.13/site-packages/autogen_agentchat/agents/_assistant_agent.py:1109: UserWarning: Resolved model mismatch: openrouter/free != arcee-ai/trinity-large-preview:free. Model mapping in autogen_ext.models.openai may be incorrect. Set the model to arcee-ai/trinity-large-preview:free to enhance token/cost estimation and suppress this warning.
  model_result = await model_client.create(
---------- TextMessage (Local_Formatter) ----------
• Precision agriculture using drone technology enhances crop monitoring through high-resolution images and multispectral data capture, enabling targeted interventions such as targeted pesticide or fertilizer applications.
• Drones significantly increase operational efficiency by allowing farmers to cover vast areas in a fraction of the time, collecting data that would be impractical to gather manually, and integrating with other precision agriculture tools.
• Data collection from drones enables data-driven decision making through real-time integration with historical information, facilitating informed decisions on planting, irrigation, harvesting, and resource optimization leading to improved productivity and profitability.